### ISSUE TITLE
[BUG] Running tests fully parallel with global setup issue with browser session state

### ISSUE THREAD

**Issue by costag1982** [microsoft/playwright#20201]
**Context:**
- Playwright Version: [1.29.0]
- Operating System: [eMac]
- Node.js version: [18.12.1 & 16.15.0]
- Browser: [Chromium]
- Extra: [React front end,GQL backend and firebase auth emulator]

**System:
    OS: macOS 13.2
    Memory: 7.13 GB / 32.00 GB
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Languages:
    Bash: 3.2.57 - /bin/bash**

This is my global setup function which essentially sets up all the data upfront and injects into the database. All the data is written to a file so it can be used in the tests. I map the the `testInfo.title` to each data group so each test has access to what it needs.

GLOBAL SETUP

```javascript
import { FullConfig } from "@playwright/test";

import setupDb from "./networkService/setupDb";

import {
  getCollectionFromScenarioData,
  directoryContainsJson,
  writeScenarioDataToFile,
  writeCacheToS3,
} from "@utils/file";
import clearDatabase from "./networkService/clearDb";

async function globalSetup(config: FullConfig): Promise<void> {
  if (!directoryContainsJson("authState")) {
    if (process.env.DELETE_TESTS != null && process.env.DELETE_TESTS === "true") {
      await clearDatabase();
    }
    await writeScenarioDataToFile();

    const collection = await getCollectionFromScenarioData();

    await setupDb(
      collection.leagueData,
      collection.userData,
      collection.playerData,
      collection.gameData,
      collection.paymentData
    );
    await writeCacheToS3();
  } else {
    console.log("authState already exists");
  }
}

export default globalSetup;

```

BASE FIXTURE

```javascript
const test = base.extend<PlayOnFixture>({
  storageState: async ({ browser }, use, testInfo) => {
    const baseUrl: string | undefined = process.env.BASE_URL;
    const currentData = await getScenarioDataFromFile(testInfo.title);

    if (currentData?.data.userData[0].email == null) {
      throw new Error("Unable to find users email from test data");
    }

    if (currentData?.data.leagueData[0].name == null) {
      throw new Error("Unable to find league name from test data");
    }

    if (baseUrl == null) {
      throw new Error("baseUrl must be set in .env");
    }

    const email = currentData?.data.userData[0].email;
    const leagueName = currentData?.data.leagueData[0].name;
    const testFileName = convertToCamelCase(testInfo.title);
    const fileName = path.join("authState", `storage-${testFileName}.json`);

    if (!directoryContainsAuthFile("authState", testFileName)) {
      const page = await browser.newPage({ storageState: undefined });
      await page.goto(baseUrl);
      await page
        .getByRole("button", { name: /get started/i })
        .first()
        .click();
      await page.getByRole("button", { name: "mail Continue with Email" }).click();
      await page.getByRole("textbox").type(email);
      await page.getByRole("button", { name: /next/i }).click();
      // eslint-disable-next-line playwright/no-wait-for-timeout
      await page.waitForTimeout(3000);
      const verificationLink = await getVerificationLink(email);

      await page.goto(`${verificationLink}`);
      await expect(page.getByRole("heading", { name: leagueName })).toBeVisible();
      await page.context().storageState({ path: fileName });
      await captureAuthFrom(page, testFileName);
      await page.close();
    } else {
      const file = fs.readFileSync(fileName, "utf-8");
      const fileData: LocalStorage = JSON.parse(file);
      const savedUserData: LocalStorageChild | undefined = fileData.origins[0].localStorage.find(
        ({ name }) => name === "userData"
      );
      if (savedUserData == null) {
        throw new Error("UserData is not set");
      }

      savedUserData.value = JSON.stringify(currentData.data.userData[0]);
      fs.writeFileSync(fileName, JSON.stringify(fileData, null, 4));
    }
    await use(fileName);
  },
```

TEST FILE

```javascript
import test from "../fixtures/basePages";

import { addAuthTo } from "@utils/firebaseAuth";

import { getScenarioDataFromFile } from "@utils/file";
import { convertToCamelCase } from "@utils/stringUtil";

test.beforeEach(async ({ page }, testInfo) => {
  await addAuthTo(page, convertToCamelCase(testInfo.title));
});

test("A player can drop out", async ({
  page,
  dashboardPage,
  availabilityComponent,
  countdownComponent,
  topNav,
}, testInfo) => {
  const { data } = await getScenarioDataFromFile(testInfo.title);
  const groupName: string = data.leagueData[0].name;
  const teamOneName: string = data.gameData?.[0].teamOneName;
  const userOneName: string = data.playerData[0].name;
  const userTwoName: string = data.playerData[1].name;

  await dashboardPage.visit();
  await topNav.expectNavTitleToBe(groupName);
  await availabilityComponent.expectGameInformationMessageToBe(
    `You are on ${teamOneName}If you need to drop out click dropout`
  );
  await countdownComponent.clickNextGameSection();
  await countdownComponent.expectPlayerToBeInTeam(userOneName);
  await countdownComponent.expectPlayerToBeInTeam(userTwoName);
  await availabilityComponent.dropoutOfGame();
  await availabilityComponent.expectGameInformationMessageToBe("Are you available to play in the next game?");
  await countdownComponent.expectPlayerToBeInTeam(userTwoName);
  await countdownComponent.expectPlayerNotToBeInTeam(userOneName);
});
```

**Describe the bug**

The issue I am having is im running all tests in parallel and consistently I get 22 tests that pass and 10 tests that fail.

However all of those 10 tests pass when I run them by themselves.

<img width="413" alt="image" src="https://user-images.githubusercontent.com/12000247/213285704-ecfbdd23-be2b-4ee6-a3c5-bb4d4a6106a5.png">

I have spent ages debugging this and I have managed to figure out that the browser has an incorrect auth token assigned to it from a leftover test

I have also attached some of the testdata file so you can see how it is set up

```json
[
    {
        "scenario": "A player can drop out",
        "data": {
            "leagueData": [
                {
                    "_id": "63c8500bcfde68f6c7cf1999",
                    "name": "Test league Watersville 17095",
                    "vip": true,
                    "gamesPlayed": 0,
                    "createdDate": "2022-12-18T20:01:15.398Z",
                    "leaguePin": "23AFR3",
                    "inviteLink": "https://playondev.page.link/4sFbREVXseDxRVUv5"
                }
            ],
            "userData": [
                {
                    "_id": "63c8500bcfde68f6c7cf1979",
                    "name": "Lavon Steuber",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "email": "montana33@example.org",
                    "provider": "email",
                    "createdDate": "2023-01-18T20:01:15.504Z",
                    "firebaseId": "9jNUx0IuEqCvhWXWjHVuvgguHSas",
                    "version": "1",
                    "leagues": [
                        {
                            "playerName": "Lavon Steuber",
                            "leagueName": "Test league Watersville 17095",
                            "leagueId": "63c8500bcfde68f6c7cf1999",
                            "playerId": "63c8500bcfde68f6c7cf199a",
                            "role": "admin",
                            "creator": true
                        }
                    ]
                },
                {
                    "_id": "63c8500bcfde68f6c7cf197a",
                    "name": "Greg Keebler",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "email": "griffin.parker@example.net",
                    "provider": "email",
                    "createdDate": "2023-01-18T20:01:15.580Z",
                    "firebaseId": "p7M3rBhXktCIM0K3OsqKxsdvHMLL",
                    "version": "1",
                    "leagues": [
                        {
                            "playerName": "Greg Keebler",
                            "leagueName": "Test league Watersville 17095",
                            "leagueId": "63c8500bcfde68f6c7cf1999",
                            "playerId": "63c8500bcfde68f6c7cf199c",
                            "role": "player",
                            "creator": false
                        }
                    ]
                }
            ],
            "playerData": [
                {
                    "_id": "63c8500bcfde68f6c7cf199a",
                    "name": "Lavon Steuber",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.400Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    },
                    "userId": "63c8500bcfde68f6c7cf1979"
                },
                {
                    "_id": "63c8500bcfde68f6c7cf199c",
                    "name": "Greg Keebler",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.400Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    },
                    "userId": "63c8500bcfde68f6c7cf197a"
                },
                {
                    "_id": "63c8500bcfde68f6c7cf199e",
                    "name": "Rachelle Keebler",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.400Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    }
                },
                {
                    "_id": "63c8500bcfde68f6c7cf19a0",
                    "name": "Danial Bayer",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.401Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    }
                },
                {
                    "_id": "63c8500bcfde68f6c7cf19a2",
                    "name": "Jamar Jerde",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.401Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    }
                },
                {
                    "_id": "63c8500bcfde68f6c7cf19a4",
                    "name": "Jamaal Bartoletti",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.401Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    }
                },
                {
                    "_id": "63c8500bcfde68f6c7cf19a6",
                    "name": "Silas Rohan",
                    "picture": "https://footy-stats-images-prod.s3.eu-west-2.amazonaws.com/5f747c83aa9d5c10d2c8ec2c-1606985761123",
                    "weeksAvailable": 0,
                    "createdDate": "2023-01-18T20:01:15.402Z",
                    "hidden": false,
                    "league": {
                        "id": "63c8500bcfde68f6c7cf1999",
                        "name": "Test league Watersville 17095"
                    }
                }
            ],
            "gameData": [
                {
                    "_id": "63c8500bcfde68f6c7cf1b5f",
                    "date": "2023-01-25T20:01:15.580Z",
                    "size": 5,
                    "leagueId": "63c8500bcfde68f6c7cf1999",
                    "result": "pending",
                    "available": [
                        "63c8500bcfde68f6c7cf199a",
                        "63c8500bcfde68f6c7cf199c",
                        "63c8500bcfde68f6c7cf199e",
                        "63c8500bcfde68f6c7cf19a0",
                        "63c8500bcfde68f6c7cf19a2",
                        "63c8500bcfde68f6c7cf19a4",
                        "63c8500bcfde68f6c7cf19a6"
                    ],
                    "notAvailable": [],
                    "teamOne": [
                        {
                            "name": "Lavon Steuber",
                            "id": "63c8500bcfde68f6c7cf199a"
                        },
                        {
                            "name": "Greg Keebler",
                            "id": "63c8500bcfde68f6c7cf199c"
                        }
                    ],
                    "teamTwo": [
                        {
                            "name": "Rachelle Keebler",
                            "id": "63c8500bcfde68f6c7cf199e"
                        },
                        {
                            "name": "Danial Bayer",
                            "id": "63c8500bcfde68f6c7cf19a0"
                        },
                        {
                            "name": "Jamar Jerde",
                            "id": "63c8500bcfde68f6c7cf19a2"
                        },
                        {
                            "name": "Jamaal Bartoletti",
                            "id": "63c8500bcfde68f6c7cf19a4"
                        },
                        {
                            "name": "Silas Rohan",
                            "id": "63c8500bcfde68f6c7cf19a6"
                        }
                    ],
                    "teamOneName": "Team One",
                    "teamTwoName": "Team Two"
                }
            ],
            "paymentData": []
        }
    },
```

I hope you can replicate this, please let me know if you need any more info

Thanks
Costa

**Comment by yury-s** [comment]
> I have spent ages debugging this and I have managed to figure out that the browser has an incorrect auth token assigned to it from a leftover test

This sounds like a problem with your tests rather than a playwright bug, if you think it's a playwright bug we need a minimal repro that we could run locally, otherwise we can't act on this report.

> The issue I am having is im running all tests in parallel and consistently I get 22 tests that pass and 10 tests that fail.
>
> However all of those 10 tests pass when I run them by themselves.

These symptoms suggest that there is a race condition between the tests. I'd start with removing write to file from the fixture and always logging in to see if it helps:

```js
...
      cons state = await page.context().storageState();
      await captureAuthFrom(page, testFileName);
      await page.close();
...
      await use(state);
```

**Comment by yury-s** [comment]
Closing as this doesn't seem to be a playwright bug, if you think otherwise please file a new issue with a repro that we can run locally and link it to this one.

**Comment by costag1982** [comment]
This issue that i have experienced is when storage state is not set. so each test is logging in via the UI from the beginning.

So it doesnt make sense how the auth tokens can get mixed up between tests.

I will do some more digging

**Comment by costag1982** [comment]
The issue is 100% in this code. As when I commented this storageState function out and logged in via each test scenario they all passed.

Everything else is the same so I still set all the data up in the global setup and write it to a file which can be read by each test

```javascript
const test = base.extend<PlayOnFixture>({
  storageState: async ({ browser }, use, testInfo) => {
    const baseUrl: string | undefined = process.env.BASE_URL;
    const currentData = await getScenarioDataFromFile(testInfo.title);

    if (currentData?.data.userData[0].email == null) {
      throw new Error("Unable to find users email from test data");
    }

    if (currentData?.data.leagueData[0].name == null) {
      throw new Error("Unable to find league name from test data");
    }

    if (baseUrl == null) {
      throw new Error("baseUrl must be set in .env");
    }

    const email = currentData?.data.userData[0].email;
    const leagueName = currentData?.data.leagueData[0].name;
    const testFileName = convertToCamelCase(testInfo.title);
    const fileName = path.join("authState", `storage-${testFileName}.json`);

    if (!directoryContainsAuthFile("authState", testFileName)) {
      const page = await browser.newPage({ storageState: undefined });
      await page.goto(baseUrl);
      await page
        .getByRole("button", { name: /get started/i })
        .first()
        .click();
      await page.getByRole("button", { name: "mail Continue with Email" }).click();
      await page.getByRole("textbox").type(email);
      await page.getByRole("button", { name: /next/i }).click();
      // eslint-disable-next-line playwright/no-wait-for-timeout
      await page.waitForTimeout(3000);
      const verificationLink = await getVerificationLink(email);

      await page.goto(`${verificationLink}`);
      await expect(page.getByRole("heading", { name: leagueName })).toBeVisible();
      await page.context().storageState({ path: fileName });
      await captureAuthFrom(page, testFileName);
      await page.close();
    } else {
      const file = fs.readFileSync(fileName, "utf-8");
      const fileData: LocalStorage = JSON.parse(file);
      const savedUserData: LocalStorageChild | undefined = fileData.origins[0].localStorage.find(
        ({ name }) => name === "userData"
      );
      if (savedUserData == null) {
        throw new Error("UserData is not set");
      }

      savedUserData.value = JSON.stringify(currentData.data.userData[0]);
      fs.writeFileSync(fileName, JSON.stringify(fileData, null, 4));
    }
    await use(fileName);
  },
```
